include/xtensor-blas/xlinalg.hpp

```diff
@@ -1514,6 +1514,172 @@ namespace linalg
+                else
+                {
+                    sum_len *= a_val;
+                }
+                ++as_it;
+                ++bs_it;
+            }
+            std::vector<std::size_t> result_shape;
+            auto r_shape_inserter = std::back_inserter(result_shape);
+            std::size_t keep_a_len = 1;
+            for (auto it = a.shape().begin(); it != a.shape().begin() + (a.dimension() - naxes); ++it)
+            {
+                std::size_t len = *it;
+                keep_a_len *= len;
+                r_shape_inserter = len;
+            }
+            std::size_t keep_b_len = 1;
+            for (auto it = b.shape().begin() + naxes; it != b.shape().end(); ++it)
+            {
+                std::size_t len = *it;
+                keep_b_len *= len;
+                r_shape_inserter = len;
+            }
+            auto a_mat = result_type::from_shape({keep_a_len, sum_len});
+            std::copy(a.storage().begin(), a.storage().end(), a_mat.storage().begin());
+            auto b_mat = result_type::from_shape({sum_len, keep_b_len});
+            std::copy(b.storage().begin(), b.storage().end(), b_mat.storage().begin());
+
+            result = dot(a_mat, b_mat);
+            result.reshape(result_shape);
+        }
+        return result;
+    }
+
+    /**
+     * @brief Compute tensor dot product along specified axes for arrays
+     *
+     * Compute the sum of products along the axes \em ax_a for a and \em ax_b for b
+     *
+     * @param xa input array
+     * @param xb input array
+     * @param ax_a axes to sum over for \em a
+     * @param ax_b axes to sum over for \em b
+     * @return resulting array
+     */
+    template <class T, class O>
+    auto tensordot(const xexpression<T>& xa, const xexpression<O>& xb, const std::vector<std::size_t>& ax_a,
+                   const std::vector<std::size_t>& ax_b)
+    {
+        const auto& a = xa.derived_cast();
+        const auto& b = xb.derived_cast();
+        XTENSOR_ASSERT(ax_a.size() == ax_b.size());
+        XTENSOR_ASSERT(ax_a.size() < a.dimension());
+        XTENSOR_ASSERT(ax_b.size() < b.dimension());
+        std::size_t n_ax = ax_a.size();
+        for (std::size_t i = 0; i < n_ax; ++i)
+        {
+            XTENSOR_ASSERT(ax_a[i] < a.dimension());
+            XTENSOR_ASSERT(ax_b[i] < b.dimension());
+        }
+
+        // Move the axes to sum over to the end of a
+        std::vector<std::size_t> newaxes_a;
+        std::vector<std::size_t> result_shape;
+        auto newax_a_inserter = std::back_inserter(newaxes_a);
+        for (std::size_t i = 0; i < a.dimension(); ++i)
+        {
+            auto a_ax_it = std::find(ax_a.begin(), ax_a.end(), i);
+            // first pass if i is not in ax_a, add to the newaxes_a, and the result shape
+            if (a_ax_it == ax_a.end())
+            {
+                newax_a_inserter = i;
+            }
+        }
+        for (auto a_ax_it = ax_a.begin(); a_ax_it != ax_a.end(); ++a_ax_it)
```

Comment:
could also use range based for loop `for (const auto& el : ax_a) ...`

Reply:
Yes, the more verbose loop structure is holdover from an implementation that was not correct. I will change this.